main.go

```diff
@@ -1,108 +1,39 @@
 package main
 
-// Great help
-// https://golangcode.com/how-to-read-a-csv-file-into-a-struct/
-
 import (
-	"encoding/csv"
-	"errors"
-	"fmt"
-	"net/http"
-	"os"
-	"strconv"
+	"log"
+
+	"rincon-orlando/go-bootcamp/config"
+	"rincon-orlando/go-bootcamp/controller"
+	"rincon-orlando/go-bootcamp/repository"
+	"rincon-orlando/go-bootcamp/service"
+	"rincon-orlando/go-bootcamp/usecase"
 
 	"github.com/gin-gonic/gin"
 )
 
-const csvFilename = "pokemons.csv"
-
-// Pokemon - Struct containing information about a Pokemon
-type Pokemon struct {
-	ID   int    `json:"id"`
-	Name string `json:"name"`
-}
-
-type DB struct {
-	data map[int]Pokemon
-}
-
 func main() {
-	pokemonDB := new(DB)
-	pokemonDB.initFromCsv(csvFilename)
-
-	// Loop through pokemons and print 'em all
-	for _, pokemon := range pokemonDB.data {
-		fmt.Printf("ID: %v, Name %v\n", pokemon.ID, pokemon.Name)
-	}
-
-	// Configure router
-	router := gin.Default()
-	router.GET("/pokemons", pokemonDB.getAllPokemons)
-	router.GET("/pokemons/:id", pokemonDB.getPokemonById)
-
-	// Start server
-	router.Run("localhost:8082")
-}
-
-// TODO: I do not like mixing DB and GIN context
-// How do I make one method for getAllPokemons from DB
-// and another one to just format as JSON?
-// i.e. how do I share context between these two objects without
-// making them dependent of each other?
-func (db *DB) getAllPokemons(c *gin.Context) {
-	// Convert DB map to array so we do not give unnecessary keys as output
-	v := make([]Pokemon, 0, len(db.data))
-	for _, value := range db.data {
-		v = append(v, value)
-	}
-	c.IndentedJSON(http.StatusOK, v)
-}
-
-func (db *DB) getPokemonById(c *gin.Context) {
-	id := c.Param("id")
-	idInt, err := strconv.Atoi(c.Param("id"))
+	config, err := config.New(".")
 	if err != nil {
-		c.IndentedJSON(http.StatusInternalServerError, gin.H{"message": "Failed to convert id " + id + " to int"})
-		return
+		log.Fatal("cannot load config:", err)
 	}
 
-	if val, ok := db.data[idInt]; ok {
-		c.IndentedJSON(http.StatusOK, val)
-		return
-	}
-
-	c.IndentedJSON(http.StatusNotFound, gin.H{"message": "pokemon with id " + id + " not found"})
-}
-
-func (db *DB) initFromCsv(filename string) (*DB, error) {
-
-	// Open CSV file
-	f, err := os.Open(filename)
+	// Dependency injection
+	db := repository.New()
+	usecase, err := usecase.New(&db, config.CSV_FILENAME)
 	if err != nil {
-		return db, err
+		log.Fatal("Error starting up database" + err.Error())
 	}
-	defer f.Close()
+	service := service.New(config.POKEMON_API_URL)
 
-	// Read file into a variable
-	lines, err := csv.NewReader(f).ReadAll()
-	if err != nil {
-		return db, err
-	}
+	controller := controller.New(usecase, service)
```

Comment:
The idea is to have the layers in this order level
router -> controller -> usecase -> service 
Is there a way you can pass the service to a function in the usecase?
To avoid the controller knowing who the service is?

Reply:
Reordering layers